8325746: Refactor Loop Unswitching code

chhagedorn opened the pull request:
To simplify the review of https://github.com/openjdk/jdk/pull/16877, I've decided to split off some code into separate sub-tasks.

This sub-task focuses on the code refactoring done in Loop Unswitching. I've incorporated the feedback from @eme64 at https://github.com/openjdk/jdk/pull/16877 in the proposed changes.

#### Changes
- Adding a more detailed description about the Loop Unswitching optimization itself below the file header.
- Renaming `fast loop` -> `true-path-loop` and `slow loop` -> `false-path-loop` since we don't really know which one is faster and to avoid keeping a mental map of which path goes into which unswitched loop version.
- Creating new classes:
  - `UnswitchedLoopSelector` to create the unswitched loop selector `If` (i.e. the If node that selects at runtime which unswitched loop version is executed) based on the invariant and non-loop-exiting unswitch candidate `If`.
  - `OriginalLoop` to do the actual unswitching (i.e. loop cloning, predicates cloning and fixing the graph to the loop selector `If`).
- Extracting methods in `do_unswitching()` to clean it up further.
- Improving `TraceLoopUnswitching` result printing.
- Adding some shared utility methods.

#### Possible Follow-Up RFEs
There are possible follow-up opportunities to further improve the loop unswitching which, however, would go to far in the context of fixing Assertion Predicates. We could follow up with RFEs to tackle these:
- At some point it might be good to have a dedicated `LoopUnswitching` class to split all the methods off from `PhaseIdealLoop`. This would also avoid noisy "loop_unswitching" prefixes.
- Bailout code in `has_control_dependencies_from_predicates()` is too conservative. I have already some code to improve that in the full fix for Assertion Predicates which I plan to split off and propose separatly.
- `TraceLoopUnswitching` currently only prints the result and if it failed. We might want to improve that in the future to cover more steps to make it useful for debugging (would require some more fine-grained debug levels, though).

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8325746](https://bugs.openjdk.org/browse/JDK-8325746): Refactor Loop Unswitching code (**Sub-task** - P4)


### Reviewers
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17842/head:pull/17842` \
`$ git checkout pull/17842`

Update a local copy of the PR: \
`$ git checkout pull/17842` \
`$ git pull https://git.openjdk.org/jdk.git pull/17842/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17842`

View PR using the GUI difftool: \
`$ git pr show -t 17842`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17842.diff">https://git.openjdk.org/jdk/pull/17842.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17842#issuecomment-1943365784)

chhagedorn commented on src/hotspot/share/opto/loopUnswitch.cpp:
```diff
@@ -102,114 +99,122 @@ IfNode* PhaseIdealLoop::find_unswitching_candidate(const IdealLoopTree *loop) co
+  }
 
-#ifndef PRODUCT
-  if (TraceLoopOpts) {
-    tty->print("Unswitch   %d ", head->unswitch_count()+1);
-    loop->dump_head();
```
Extracted to `trace_loop_unswitching_count()`

chhagedorn commented on src/hotspot/share/opto/loopUnswitch.cpp:
```diff
@@ -102,114 +99,122 @@ IfNode* PhaseIdealLoop::find_unswitching_candidate(const IdealLoopTree *loop) co
+  }
 
-  // Need to revert back to normal loop
-  if (head->is_CountedLoop() && !head->as_CountedLoop()->is_normal_loop()) {
-    head->as_CountedLoop()->set_normal_loop();
```
Extracted to `revert_to_normal_loop()`

chhagedorn commented on src/hotspot/share/opto/loopUnswitch.cpp:
```diff
@@ -102,114 +99,122 @@ IfNode* PhaseIdealLoop::find_unswitching_candidate(const IdealLoopTree *loop) co
+  if (has_control_dependencies_from_predicates(original_head)) {
+    NOT_PRODUCT(trace_loop_unswitching_impossible(original_head);)
+    return;
   }
```
Extracted to `hoist_invariant_check_casts()`

chhagedorn commented on src/hotspot/share/opto/loopUnswitch.cpp:
```diff
@@ -102,114 +99,122 @@ IfNode* PhaseIdealLoop::find_unswitching_candidate(const IdealLoopTree *loop) co
+  }
 
-  // Increment unswitch count
-  LoopNode* head_clone = old_new[head->_idx]->as_Loop();
-  int nct = head->unswitch_count() + 1;
-  head->set_unswitch_count(nct);
-  head_clone->set_unswitch_count(nct);
```
Extracted to `increment_unswitch_counts()`.

chhagedorn commented on src/hotspot/share/opto/loopUnswitch.cpp:
```diff
@@ -102,114 +99,122 @@ IfNode* PhaseIdealLoop::find_unswitching_candidate(const IdealLoopTree *loop) co
-  ProjNode* proj_false = invar_iff->proj_out(0);
-  dominated_by(proj_false->as_IfProj(), unswitch_iff_clone);
-
-  // Reoptimize loops
-  loop->record_for_igvn();
-  for(int i = loop->_body.size() - 1; i >= 0 ; i--) {
-    Node *n = loop->_body[i];
-    Node *n_clone = old_new[n->_idx];
-    _igvn._worklist.push(n_clone);
-  }
```
Extracted to `add_unswitched_loop_version_bodies_to_igvn()`

chhagedorn commented on src/hotspot/share/opto/loopUnswitch.cpp:
```diff
@@ -102,114 +99,122 @@ IfNode* PhaseIdealLoop::find_unswitching_candidate(const IdealLoopTree *loop) co
-  }
+  revert_to_normal_loop(original_head);
 
-#ifndef PRODUCT
-  if (TraceLoopUnswitching) {
-    tty->print_cr("Loop unswitching orig: %d @ %d  new: %d @ %d",
-                  head->_idx,                unswitch_iff->_idx,
-                  old_new[head->_idx]->_idx, unswitch_iff_clone->_idx);
-  }
-#endif
```
Extracted to `trace_loop_unswitching_result()`

chhagedorn commented on src/hotspot/share/opto/loopUnswitch.cpp:
```diff
@@ -102,114 +99,122 @@ IfNode* PhaseIdealLoop::find_unswitching_candidate(const IdealLoopTree *loop) co
+    return _selector;
   }
 
-  IfNode* invar_iff = create_slow_version_of_loop(loop, old_new, unswitch_iff, CloneIncludesStripMined);
-  ProjNode* proj_true = invar_iff->proj_out(1);
-  verify_fast_loop(head, proj_true);
```
Refactored to `create_unswitched_loop_versions()`.

chhagedorn commented on src/hotspot/share/opto/loopUnswitch.cpp:
```diff
@@ -102,114 +99,122 @@ IfNode* PhaseIdealLoop::find_unswitching_candidate(const IdealLoopTree *loop) co
-  // Hardwire the control paths in the loops into if(true) and if(false)
-  _igvn.rehash_node_delayed(unswitch_iff);
-  dominated_by(proj_true->as_IfProj(), unswitch_iff);
+  NOT_PRODUCT(trace_loop_unswitching_count(loop, original_head);)
+  C->print_method(PHASE_BEFORE_LOOP_UNSWITCHING, 4, original_head);
 
-  IfNode* unswitch_iff_clone = old_new[unswitch_iff->_idx]->as_If();
-  _igvn.rehash_node_delayed(unswitch_iff_clone);
-  ProjNode* proj_false = invar_iff->proj_out(0);
-  dominated_by(proj_false->as_IfProj(), unswitch_iff_clone);
```
Extracted to `remove_unswitch_candidate_from_loops()` and put inside `create_unswitched_loop_versions()`.

chhagedorn commented on src/hotspot/share/opto/loopopts.cpp:
```diff
@@ -2867,8 +2867,7 @@ ProjNode* PhaseIdealLoop::insert_if_before_proj(Node* left, bool Signed, BoolTes
 
   int opcode = iff->Opcode();
   assert(opcode == Op_If || opcode == Op_RangeCheck, "unexpected opcode");
-  IfNode* new_if = (opcode == Op_If) ? new IfNode(proj2, bol, iff->_prob, iff->_fcnt):
-    new RangeCheckNode(proj2, bol, iff->_prob, iff->_fcnt);
```
Since I re-used this pattern again, I extracted this into `make_with_same_profile()`.

eme64 replied:
Do you want to take over the assert as well?

chhagedorn replied:
I think it should now be covered with the additional assert in `make_with_same_profile()`. If we pass something else, we will hit that assert.

eme64 commented on src/hotspot/share/opto/loopUnswitch.cpp:
```diff
@@ -102,114 +99,122 @@ IfNode* PhaseIdealLoop::find_unswitching_candidate(const IdealLoopTree *loop) co
+ private:
+  IfNode* find_unswitch_candidate(IdealLoopTree* loop) {
+    IfNode* unswitch_candidate = _phase->find_unswitch_candidate(loop);
+    assert(unswitch_candidate != nullptr, "must exist");
```
```suggestion
    assert(unswitch_candidate != nullptr, "guaranteed to exist by policy_unswitching");
```

eme64 commented on src/hotspot/share/opto/loopUnswitch.cpp:
```diff
@@ -102,114 +99,122 @@ IfNode* PhaseIdealLoop::find_unswitching_candidate(const IdealLoopTree *loop) co
-    loop->dump_head();
+  IfProjNode* create_proj_to_loop(const PathToLoop path_to_loop) {
+    const uint dom_depth = _phase->dom_depth(_original_loop_entry);
+    IfProjNode* proj_to_loop;
+    if (path_to_loop == TRUE_PATH) {
+      proj_to_loop = new IfTrueNode(_selector);
+    } else {
+      proj_to_loop = new IfFalseNode(_selector);
+    }
```
```suggestion
    IfProjNode* proj_to_loop = (path_to_loop == TRUE_PATH) ? new IfTrueNode(_selector)
                                                             new IfFalseNode(_selector);
```
A question of taste. Just an idea.

eme64 reviewed (changes requested):
Nice work! A first batch of comments.

chhagedorn commented on src/hotspot/share/opto/loopUnswitch.cpp:
```diff
@@ -31,31 +31,30 @@
 #include "opto/rootnode.hpp"
 
-//================= Loop Unswitching =====================
+// Loop Unswitching is a loop optimization to move an invariant, non-loop-exiting test in the loop body before the loop.
+// This is done by duplicating the loop and placing the if-path in one unswitched loop version (i.e. the true-path-loop)
+// and the else-path in the other (i.e. the false-path-loop). The invariant test only needs to be checked once and
+// depending on the outcome, one or the other unswitched loop version is executed:
```
> BTW: why do we not hoist exit conditions also? Would that not also be profitable? Just out of curiosity...

This can better be done with a simple loop peeling which can hoist the invariant checks. Also see:
https://github.com/openjdk/jdk/blob/0c2def0e3eaa4925746158294f798d37d42c50f3/src/hotspot/share/opto/loopTransform.cpp#L499-L502

eme64 replied:
Ah yes, makes sense!
Maybe add a small comment about loop peeling then? Just because you do mention that we don't handle exits, that automatically raises that question.

eme64 commented on src/hotspot/share/opto/loopUnswitch.cpp:
```diff
@@ -222,69 +227,172 @@ bool PhaseIdealLoop::has_control_dependencies_from_predicates(LoopNode* head) co
+void PhaseIdealLoop::add_unswitched_loop_version_bodies_to_igvn(IdealLoopTree* loop, const Node_List& old_new) {
+  loop->record_for_igvn();
+  for(int i = loop->_body.size() - 1; i >= 0 ; i--) {
+    Node *n = loop->_body[i];
+    Node *n_clone = old_new[n->_idx];
```
```suggestion
    Node* n = loop->_body[i];
    Node* n_clone = old_new[n->_idx];
```

eme64 commented on src/hotspot/share/opto/loopUnswitch.cpp:
```diff
@@ -222,69 +227,172 @@ bool PhaseIdealLoop::has_control_dependencies_from_predicates(LoopNode* head) co
+  }
+};
+
+// Create a true-path- and a false-path-loop version of the original loop by cloning the loop and inserting an If to
```
Don't you already have the selector if at this point? The comment suggests you are inserting it here.

eme64 commented on src/hotspot/share/opto/loopUnswitch.cpp:
```diff
@@ -222,69 +227,172 @@ bool PhaseIdealLoop::has_control_dependencies_from_predicates(LoopNode* head) co
 
+ public:
+  // Unswitch on the invariant loop selector by creating two unswitched loop versions.
+  void unswitch(const UnswitchedLoopSelector& unswitched_loop_selector) {
```
Just a suggestion, I am also ok with what it is now:
I wonder if the name `unswitch` is right here. Because at this point you only clone the loops, you don't yet remove the if. Maybe this should for now just be called "duplicate", and `create_unswitched_loop_versions` then really does the unswitching, by both duplicating and removing the ifs?

eme64 commented on src/hotspot/share/opto/loopUnswitch.cpp:
```diff
@@ -222,69 +227,172 @@ bool PhaseIdealLoop::has_control_dependencies_from_predicates(LoopNode* head) co
+class OriginalLoop : public StackObj {
+  LoopNode* const _strip_mined_loop_head;
+  IdealLoopTree* const _loop;
+  Node_List* const _old_new;
```
Could also be held "by reference". Then you don't have convert it from and to pointer. Well, idk, maybe it would be just as bad the other way around.

eme64 commented on src/hotspot/share/opto/loopUnswitch.cpp:
```diff
@@ -222,69 +227,172 @@ bool PhaseIdealLoop::has_control_dependencies_from_predicates(LoopNode* head) co
+#endif
 
-  return iff;
+void PhaseIdealLoop::revert_to_normal_loop(const LoopNode* loop_head) {
+  if (loop_head->is_CountedLoop() && !loop_head->as_CountedLoop()->is_normal_loop()) {
+    loop_head->as_CountedLoop()->set_normal_loop();
+  }
```
```suggestion
void PhaseIdealLoop::revert_to_normal_loop(const LoopNode* loop_head) {
  CountedLoopNode* cl = loop_head->isa_CountedLoop();
  if (cl != nullptr && !cl->is_normal_loop()) {
    cl->set_normal_loop();
  }
```

eme64 replied:
Would make the code easier to read

chhagedorn replied:
I just moved the code but I can clean it up as you suggested.

eme64 reviewed (changes requested):
Nice Work @chhagedorn . I like that you packed things into classes. It nicely packs things into more digestable concepts. And allows you to create objects of those classes, and then pass then around as `const`, after which the reader knows they are not changed anymore. Very helpful!

chhagedorn commented on src/hotspot/share/opto/loopUnswitch.cpp:
```diff
@@ -222,69 +227,172 @@ bool PhaseIdealLoop::has_control_dependencies_from_predicates(LoopNode* head) co
-  assert(proj_true->is_IfTrue(), "must be true projection");
-  Node* entry = head->skip_strip_mined()->in(LoopNode::EntryControl);
-  Predicates predicates(entry);
-  if (!predicates.has_any()) {
-    // No Parse Predicate.
-    Node* uniqc = proj_true->unique_ctrl_out();
-    assert((uniqc == head && !head->is_strip_mined()) || (uniqc == head->in(LoopNode::EntryControl)
-                                                          && head->is_strip_mined()), "must hold by construction if no predicates");
```
This is now covered by 

```
    const Predicates predicates(entry);
    // When skipping all predicates, we should end up at 'loop_selector_if_proj'.
    assert(loop_selector_if_proj == predicates.entry(), "should end up at loop selector If");
```

chhagedorn commented on src/hotspot/share/opto/loopUnswitch.cpp:
```diff
@@ -222,69 +227,172 @@ bool PhaseIdealLoop::has_control_dependencies_from_predicates(LoopNode* head) co
+                                                 const UnswitchedLoopSelector& unswitched_loop_selector) {
+  IfNode* unswitch_candidate = unswitched_loop_selector.unswitch_candidate();
+  IfNode* loop_selector = unswitched_loop_selector.selector();
+  Node_List worklist;
```
I think the code of this entire method is somewhat confusing. I've just moved it and did not try to refactor this code. I can add the resource mark and do the suggested renaming/retyping.

chhagedorn reviewed (commented):
Thanks a lot for your suggestions! I've pushed updates addressing them all.

eme64 commented on src/hotspot/share/opto/loopUnswitch.cpp:
```diff
@@ -24,38 +24,54 @@
+//        if-block                    =========>              [Cloned Template                  [Cloned Template
+//      else                                                   Assertion Predicates]             Assertion Predicates]
+//        // could be empty                                         |                                  |
+//        [else-block]                                        True-Path-Loop                    False-Path-Loop
```
Could it also be the other way around, that the if-path is empty but the else-path has code?

chhagedorn replied:
Yes definitely, but this is more a reflection of Java code where you always have an if block but not always an else block. But it might be cleaner to just remove the "could be empty" comment and make it explicit and not worry about whether one block is empty. I'll do that. The image should just give a hint :-)

eme64 commented on src/hotspot/share/opto/loopUnswitch.cpp:
```diff
@@ -24,38 +24,54 @@
+// Loop Unswitching is a loop optimization to move an invariant, non-loop-exiting test in the loop body before the loop.
+// Such a test is either always true or always false in all loop iterations and could therefore only be executed once.
+// To achieve that, we duplicate the loop and change the original and cloned loop as follows:
+// - Original loop -> true-path-loop: The true-path-path of the invariant, non-loop-exiting test in the original loop
+//                                    is kept while the false-path-path is killed. We call this unswitched loop version
+//                                    the true-path-loop.
+// - Cloned loop -> False-path-loop:  The false-path-path of the invariant, non-loop-exiting test in the cloned loop
+//                                    is kept while the true-path-path is killed. We call this unswitched loop version
+//                                    the false-path loop.
```
```suggestion
// - Original loop -> true-path-loop: The true-path of the invariant, non-loop-exiting test in the original loop
//                                    is kept while the false-path is killed. We call this unswitched loop version
//                                    the true-path-loop.
// - Cloned loop -> False-path-loop:  The false-path of the invariant, non-loop-exiting test in the cloned loop
//                                    is kept while the true-path is killed. We call this unswitched loop version
//                                    the false-path loop.
```

eme64 commented on src/hotspot/share/opto/loopUnswitch.cpp:
```diff
@@ -24,38 +24,54 @@
+//      stmt2                                                   cloned stmt1                      cloned stmt1
+//    Endloop                                                   cloned if-path                    [cloned else-path]
+//                                                              cloned stmt2                      cloned stmt2
+//                                                            Endloop                           Endloop
```
Could it be the other way around: the if-path is empty and the else-path has code? Or are you just trying to suggest that there could be an empty path, and that this would really simplify one of the cloned loops now?

chhagedorn replied:
Yes, it would be a simplification. But as suggested in the other comment. It might create more confusion than actually helping.

eme64 commented on src/hotspot/share/opto/loopUnswitch.cpp:
```diff
@@ -102,114 +116,204 @@ IfNode* PhaseIdealLoop::find_unswitching_candidate(const IdealLoopTree *loop) co
+
+ public:
+  OriginalLoop(IdealLoopTree* loop, Node_List& old_new)
+      : _strip_mined_loop_head(loop->_head->as_Loop()->skip_strip_mined()),
```
Hmm. this could now be a strip mined loop node, or a regular loop node (if loop not strip mined), right?
Maybe you could just call it `_loop_head`, and leave a comment, see suggestion above.

eme64 commented on src/hotspot/share/opto/loopUnswitch.cpp:
```diff
@@ -102,114 +116,204 @@ IfNode* PhaseIdealLoop::find_unswitching_candidate(const IdealLoopTree *loop) co
+// Class to unswitch the original loop and create Predicates at the new unswitched loop versions. The newly cloned loop
+// becomes the false-path-loop while original loop becomes the true-path-loop.
+class OriginalLoop : public StackObj {
+  LoopNode* const _strip_mined_loop_head;
```
```suggestion
  LoopNode* const _loop_head; // OuterStripMinedLoopNode if loop strip mined, else just the loop head.
```

eme64 reviewed (changes requested):
Thanks for the updates, we are almost there :)

vnkozlov commented on src/hotspot/share/opto/loopnode.hpp:
```diff
@@ -770,6 +771,11 @@ class IdealLoopTree : public ResourceObj {
     return _has_range_checks;
   }
 
+  // Return the strip mined loop tree if there is an outer strip mined loop. Otherwise, return this.
```
The comment is not clear for me. Something like next?:
"// Return parent's IdealLoopTree for strip mined loop and `this` for other cases."

vnkozlov commented on src/hotspot/share/opto/loopUnswitch.cpp:
```diff
@@ -24,38 +24,54 @@
+// Loop Unswitching is a loop optimization to move an invariant, non-loop-exiting test in the loop body before the loop.
+// Such a test is either always true or always false in all loop iterations and could therefore only be executed once.
+// To achieve that, we duplicate the loop and change the original and cloned loop as follows:
+// - Original loop -> true-path-loop: The true-path of the invariant, non-loop-exiting test in the original loop
+//                                    is kept while the false-path is killed. We call this unswitched loop version
+//                                    the true-path-loop.
+// - Cloned loop -> False-path-loop:  The false-path of the invariant, non-loop-exiting test in the cloned loop
+//                                    is kept while the true-path is killed. We call this unswitched loop version
+//                                    the false-path loop.
```
Comments are too wide and difficult to read. Suggestion is to put explanation on new line with shift:
```
// - Original loop -> true-path-loop:
//        The true-path of the invariant, non-loop-exiting test in the original loop
//        is kept while the false-path is killed. We call this unswitched loop version
//        the true-path-loop.
```

vnkozlov commented on src/hotspot/share/opto/loopUnswitch.cpp:
```diff
@@ -24,38 +24,54 @@
+//    Endloop                                                   cloned stmt1                      cloned stmt1
+//                                                              cloned if-path                    cloned else-path
+//                                                              cloned stmt2                      cloned stmt2
+//                                                            Endloop                           Endloop
```
Can you put resulting unstitched graph on separate following lines? Comment line are too long.

vnkozlov reviewed (commented):
Good refactoring. My main complain is long comments which are difficult to read if you edit file in terminal.

chhagedorn commented:
Thanks @eme64 and @vnkozlov for your reviews! I've addressed your latest comments.

About the long lines: I'm using 120 chars by default which fits nicely on the screen in the IDE and is also the default suggested by the IDE. Unfortunately, our style guide also does not mention anything about it - maybe also because it's personal preference. But anyway, I agree that some comments were a bit too densely packed. I've updated them according to your suggestions. Let me know if there are other comments that you think should be updated as well.

vnkozlov reviewed (approved):
Much better.

eme64 commented on src/hotspot/share/opto/loopUnswitch.cpp:
```diff
@@ -102,114 +134,204 @@ IfNode* PhaseIdealLoop::find_unswitching_candidate(const IdealLoopTree *loop) co
-  // Hardwire the control paths in the loops into if(true) and if(false)
-  _igvn.rehash_node_delayed(unswitch_iff);
-  dominated_by(proj_true->as_IfProj(), unswitch_iff);
+// Remove the unswitch candidate If nodes in both unswitched loop versions which are now dominated by the loop selector
+// If node. Keep the true-path-path in the true-path-loop and the false-path-path in the false-path-loop by setting
+// the bool input accordingly. The unswitch candidate If nodes are folded in the next IGVN round.
```
```suggestion
  // Remove the unswitch candidate If nodes in both unswitched loop versions which are now dominated by the loop selector
  // If node. Keep the true-path-path in the true-path-loop and the false-path-path in the false-path-loop by setting
  // the bool input accordingly. The unswitch candidate If nodes are folded in the next IGVN round.
```

eme64 reviewed (approved):
Nice work, Christian!

chhagedorn commented:
Thanks Vladimir and Emanuel for your reviews and also Emanuel for the offline discussion!

chhagedorn commented:
I'll resubmit some testing again before integration.

chhagedorn commented:
Testing showed that `make_with_same_profile()` could also be called with a `BaseCountedEndLoopNode` (attached extracted test for int and long). We could be unswitching on a `BaseCountedEndLoopNode` after peeling one iteration - the exit test then serves as a zero trip guard. I've changed the code accordingly to fall back to `IfNode` if it's not a `RangeCheckNode`. I've also changed the assertion to allow all these nodes but not a `ParsePredicateNode` on which `make_with_same_profile()` should never be called.

eme64 commented on test/hotspot/jtreg/compiler/loopopts/TestBaseCountedEndLoopUnswitchCandidate.java:
```diff
@@ -0,0 +1,100 @@
+ * @requires vm.compiler2.enabled
+ * @run main/othervm -XX:CompileCommand=compileonly,compiler.loopopts.TestBaseCountedEndLoopUnswitchCandidate::test*
+ *                   -Xcomp -XX:LoopMaxUnroll=0 -XX:-UseLoopPredicate -XX:-RangeCheckElimination
+ *                   compiler.loopopts.TestBaseCountedEndLoopUnswitchCandidate
```
Might it make sense to have a run with fewer flags?

chhagedorn replied:
Do you mean an additional run? For the current run, we require the C2 flags in order to have loop peeling -> loop unswitching to get a `BaseCountedLoopEndNode`. We might be able to find a test that requires fewer flags but I think it's okay to opt for a simpler test + flags.

chhagedorn replied:
Sounds good, added such a run.

eme64 commented on src/hotspot/share/opto/ifnode.cpp:
```diff
@@ -455,11 +455,17 @@ static Node* split_if(IfNode *iff, PhaseIterGVN *igvn) {
+    // the same profiling if that actually makes sense. Some If node subtypes should not be cloned in this way.
+    // In theory, we should not clone BaseCountedLoopEndNodes. But they can end up being used as normal If nodes when
+    // peeling a loop - they serve as zero-trip guard. Allow them as well.
+    assert(if_node_profile->Opcode() == Op_If || if_node_profile->is_RangeCheck()
```
`if_node_profile->is_RangeCheck()` seems excluded by the if-else above...
You could move the assert before the if, then it would be correct, I think

eme64 commented on test/hotspot/jtreg/compiler/loopopts/TestBaseCountedEndLoopUnswitchCandidate.java:
```diff
@@ -0,0 +1,100 @@
+    }
+
+
+    public static void testLongCountedLoopEnd() {
```
Remove a newline

eme64 reviewed (approved):
Update makes sense. A few suggestions below.

chhagedorn reviewed (commented):
Thanks Emanuel for your re-review! Pushed an update.

eme64 commented on test/hotspot/jtreg/compiler/loopopts/TestBaseCountedEndLoopUnswitchCandidate.java:
```diff
@@ -0,0 +1,100 @@
+ * @run main/othervm -XX:CompileCommand=compileonly,compiler.loopopts.TestBaseCountedEndLoopUnswitchCandidate::test*
+ *                   -Xcomp -XX:LoopMaxUnroll=0 -XX:-UseLoopPredicate -XX:-RangeCheckElimination
+ *                   compiler.loopopts.TestBaseCountedEndLoopUnswitchCandidate
+ * @run main/othervm compiler.loopopts.TestBaseCountedEndLoopUnswitchCandidate
```
```suggestion
 * @run driver compiler.loopopts.TestBaseCountedEndLoopUnswitchCandidate
```
Nit: now that you have no arguments

chhagedorn replied:
As discussed offline, changing this to `@run main` to enable the CI to pass additional flags which would not otherwise be possible with `driver`.

vnkozlov reviewed (approved):
This version looks good.

chhagedorn commented:
Thanks again @eme64 and @vnkozlov for your reviews and comments!

/integrate
